fix: correct export/import of OpenAI Compatible codebase indexing set…

MuriloFP opened the pull request:
# fix: correct export/import of OpenAI Compatible codebase indexing settings (#5035)

<!--
Thank you for contributing to Roo Code!

Before submitting your PR, please ensure:
- It's linked to an approved GitHub Issue.
- You've reviewed our [Contributing Guidelines](../CONTRIBUTING.md).
-->

### Related GitHub Issue

<!-- Every PR MUST be linked to an approved issue. -->

Closes: #5035 <!-- Replace with the issue number, e.g., Closes: #123 -->

### Roo Code Task Context (Optional)

<!--
If you used Roo Code to help create this PR, you can share public task links here.
This helps reviewers understand your development process and provides additional context.
Example: https://app.roocode.com/share/task-id
-->

### Description

<!--
Briefly summarize the changes in this PR and how they address the linked issue.
The issue should cover the "what" and "why"; this section should focus on:
- The "how": key implementation details, design choices, or trade-offs made.
- Anything specific reviewers should pay attention to in this PR.
-->

This PR fixes the export/import functionality for OpenAI Compatible codebase indexing settings. The issue was that exported settings were missing the `codebaseIndexEmbedderModelDimension` property and contained an incorrect `codebaseIndexEmbedderBaseUrl`.

**Key implementation details:**
- Modified `exportSettings` function to correctly fetch OpenAI Compatible settings from global state instead of provider profiles
- Added missing `codebaseIndexEmbedderModelDimension` property to the `codebaseIndexConfigSchema` type definition
- Updated test mocking strategy to properly mock `contextProxy.getGlobalState` calls

**Design choice:** The fix maintains the existing architecture where OpenAI Compatible indexing settings are stored in global state rather than provider profiles, ensuring consistency with the current implementation.

### Test Procedure

<!--
Detail the steps to test your changes. This helps reviewers verify your work.
- How did you test this specific implementation? (e.g., unit tests, manual testing steps)
- How can reviewers reproduce your tests or verify the fix/feature?
- Include relevant testing environment details if applicable.
-->

**Automated Testing:**
- Added new test case specifically for OpenAI Compatible export functionality
- Updated 8 existing tests to align with the corrected implementation
- All 2643 tests pass successfully

**Manual Testing Steps:**
1. Configure OpenAI Compatible provider for codebase indexing with custom base URL and model dimension
2. Export settings via Settings > Import/Export > Export Settings
3. Verify exported JSON contains both `codebaseIndexEmbedderBaseUrl` and `codebaseIndexEmbedderModelDimension`
4. Import the exported settings on a fresh instance
5. Verify all OpenAI Compatible settings are correctly restored

**Test Commands:**
```bash
cd src
npx vitest core/config/__tests__/importExport.spec.ts

### Pre-Submission Checklist

<!-- Go through this checklist before marking your PR as ready for review. -->

- [X] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [X] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [X] **Self-Review**: I have performed a thorough self-review of my code.
- [X] **Testing**: New and/or updated tests have been added to cover my changes (if applicable).
- [X] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [X] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

### Screenshots / Videos

<!--
For UI changes, please provide before-and-after screenshots or a short video of the *actual results*.
This greatly helps in understanding the visual impact of your changes.
-->

### Documentation Updates

<!--
Does this PR necessitate updates to user-facing documentation?
- [X] No documentation updates are required.
- [ ] Yes, documentation updates are required. (Please describe what needs to be updated or link to a PR in the docs repository).
-->

### Additional Notes

<!-- Add any other context, questions, or information for reviewers here. -->
This fix ensures that users can properly export and import their OpenAI Compatible codebase indexing configurations, including custom base URLs and model dimensions. The bug was caused by the export function looking for these settings in the wrong location (provider profiles instead of global state).

I had some extra issues with this PR because the import/export settings were not properly updating the UI state and the OpenAI Compatible Provider specific fields... The tests were created to ensure that these work correctly.

### Get in Touch

<!--
Please provide your Discord username for reviewers or maintainers to reach you if they have questions about your PR
-->
@MuriloFP

daniel-lxs reviewed (commented):
I've conducted a thorough analysis of this PR and discovered a critical architectural issue that explains the root cause of all three linked issues (#5035, #4962, #5091).

## Critical Finding: Storage Location Mismatch

There's a fundamental mismatch between where OpenAI Compatible settings are saved vs. where they're read:

### 1. Writing (src/core/webview/webviewMessageHandler.ts lines 1841-1842)
Settings are saved INSIDE the `codebaseIndexConfig` object:
```typescript
const globalStateConfig = {
  ...currentConfig,
  codebaseIndexOpenAiCompatibleBaseUrl: settings.codebaseIndexOpenAiCompatibleBaseUrl,
  codebaseIndexOpenAiCompatibleModelDimension: settings.codebaseIndexOpenAiCompatibleModelDimension,
}
await updateGlobalState("codebaseIndexConfig", globalStateConfig)
```

### 2. Reading (src/services/code-index/config-manager.ts lines 62-65)
The code tries to read them as SEPARATE global state keys:
```typescript
const openAiCompatibleBaseUrl = this.contextProxy?.getGlobalState("codebaseIndexOpenAiCompatibleBaseUrl") ?? ""
const openAiCompatibleModelDimension = this.contextProxy?.getGlobalState("codebaseIndexOpenAiCompatibleModelDimension")
```

## Why This Causes All Three Issues

- **#5035**: Export reads from the wrong location (separate keys that don't exist)
- **#4962**: Settings "disappear" because they're saved in one place but read from another
- **#5091**: Indexing fails to start because config-manager can't find the settings

## Current PR Assessment

While this PR does fix the export/import functionality, it's essentially implementing a workaround by:
1. Reading from global state during export (which happens to work because it reads from the correct location)
2. Duplicating settings to provider profiles during import

The complex provider mapping logic (lines 72-112) is trying to work around the core issue rather than fixing it.

## Recommended Solution

Instead of this workaround, we should fix the root cause. Two options:

**Option 1**: Update config-manager.ts to read from the correct location:
```typescript
const codebaseIndexConfig = this.contextProxy?.getGlobalState("codebaseIndexConfig") ?? {}
const openAiCompatibleBaseUrl = codebaseIndexConfig.codebaseIndexOpenAiCompatibleBaseUrl ?? ""
const openAiCompatibleModelDimension = codebaseIndexConfig.codebaseIndexOpenAiCompatibleModelDimension
```

**Option 2**: Update webviewMessageHandler.ts to save as separate keys:
```typescript
await updateGlobalState("codebaseIndexOpenAiCompatibleBaseUrl", settings.codebaseIndexOpenAiCompatibleBaseUrl)
await updateGlobalState("codebaseIndexOpenAiCompatibleModelDimension", settings.codebaseIndexOpenAiCompatibleModelDimension)
```

## Code Quality Issues

1. **Unused variable** (line 77): `targetProviderName` is assigned but never used
2. **Empty catch block** (line 227): Silently swallows errors without logging

## Questions for Consideration

1. Should we fix the root cause instead of implementing this workaround?
2. If we keep the provider profile approach, what's the migration strategy for existing users?
3. Why maintain duplicate state in both global state and provider profiles?

While this PR does solve the immediate symptoms, addressing the underlying storage mismatch would provide a cleaner, more maintainable solution.

MuriloFP commented:
I've addressed the root cause issue identified in the review. The fix includes:

1. **Fixed the storage mismatch in config-manager.ts**: The OpenAI Compatible settings are now correctly read from the nested \codebaseIndexConfig\ object instead of non-existent top-level keys.

2. **Updated type definitions**: Added the OpenAI Compatible fields to the \CodebaseIndexConfig\ interface to ensure type safety.

3. **Removed workaround code**: The complex logic in import/export that was compensating for the bug is no longer needed and has been removed.

4. **Fixed UI display bug**: The settings popover now correctly displays saved OpenAI Compatible settings instead of showing blank fields.

5. **Updated tests**: All unit tests have been updated to match the corrected architecture.

The settings now properly persist across restarts, display correctly in the UI, and are included in exports/imports as expected. All tests are passing.

MuriloFP commented:
<!--
Thank you for contributing to Roo Code!

Before submitting your PR, please ensure:
- It's linked to an approved GitHub Issue.
- You've reviewed our [Contributing Guidelines](../CONTRIBUTING.md).
-->

### Related GitHub Issue

<!-- Every PR MUST be linked to an approved issue. -->

Closes: #5035
Also fixes: #4962, #5091

### Roo Code Task Context (Optional)

<!--
If you used Roo Code to help create this PR, you can share public task links here.
This helps reviewers understand your development process and provides additional context.
Example: https://app.roocode.com/share/task-id
-->

_No Roo Code task context for this PR._

### Description

<!--
Briefly summarize the changes in this PR and how they address the linked issue.
The issue should cover the "what" and "why"; this section should focus on:
- The "how": key implementation details, design choices, or trade-offs made.
- Anything specific reviewers should pay attention to in this PR.
-->

This PR addresses the review feedback and fixes identified issues for PR #5383.

**Key Changes:**
- Fixed the root cause of OpenAI Compatible codebase indexing settings persistence issues
- Resolved merge conflicts with the latest main branch by properly integrating both `codebaseIndexSearchMaxResults` and OpenAI Compatible fields
- Ensured proper alignment between storage write and read operations for OpenAI Compatible settings
- All settings now correctly persist within the `codebaseIndexConfig` object structure

**Review Comments Addressed:**
- ✅ Code quality issues (unused variable and empty catch block) - Already resolved during rebase
- ✅ Merge conflicts in `packages/types/src/codebase-index.ts` - Successfully resolved
- ✅ Root cause fix verified - Storage location mismatch properly addressed

**Root Cause Analysis:**
The core issue was a mismatch between where settings were written (inside `codebaseIndexConfig`) and where they were read (from non-existent top-level keys). This has been fixed by ensuring both read and write operations use the same nested structure.

### Test Procedure

<!--
Detail the steps to test your changes. This helps reviewers verify your work.
- How did you test this specific implementation? (e.g., unit tests, manual testing steps)
- How can reviewers reproduce your tests or verify the fix/feature?
- Include relevant testing environment details if applicable.
-->

**Testing performed:**
1. Ran all unit tests locally: `npm test`
   - `importExport.spec.ts`: 32/32 tests passed
   - `config-manager.spec.ts`: 51/51 tests passed
   - `service-factory.spec.ts`: 26/26 tests passed
2. Ran TypeScript type checking: `npx tsc --noEmit`
3. Manual testing steps:
   - Configure OpenAI Compatible provider in Code Index settings
   - Export settings and verify OpenAI Compatible fields are included
   - Restart VSCode and verify settings persist
   - Import settings and verify OpenAI Compatible fields are restored

**To verify these changes:**
1. Check out this branch
2. Run `npm install && npm test`
3. Configure an OpenAI Compatible provider in the Code Index settings
4. Export settings and verify the JSON includes `codebaseIndexOpenAiCompatibleBaseUrl` and `codebaseIndexOpenAiCompatibleModelDimension`
5. Restart VSCode and confirm settings persist

**Test Environment:**
- Node.js version: 18.x
- OS: Windows 11
- VSCode version: Latest

### Pre-Submission Checklist

<!-- Go through this checklist before marking your PR as ready for review. -->

- [x] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [x] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [x] **Self-Review**: I have performed a thorough self-review of my code.
- [x] **Testing**: New and/or updated tests have been added to cover my changes (if applicable).
- [x] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [x] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

### Screenshots / Videos

<!--
For UI changes, please provide before-and-after screenshots or a short video of the *actual results*.
This greatly helps in understanding the visual impact of your changes.
-->

_No UI changes in this PR._

### Documentation Updates

<!--
Does this PR necessitate updates to user-facing documentation?
- [ ] No documentation updates are required.
- [ ] Yes, documentation updates are required. (Please describe what needs to be updated or link to a PR in the docs repository).
-->

- [x] No documentation updates are required.

### Additional Notes

<!-- Add any other context, questions, or information for reviewers here. -->

All review feedback has been addressed. The rebase process resolved the code quality issues mentioned in the review, and the merge conflicts have been properly resolved. The fix addresses the root cause of three related issues (#5035, #4962, #5091) by ensuring consistent storage location for OpenAI Compatible settings.

**Files Modified:**
```
packages/types/src/codebase-index.ts - Resolved merge conflict, integrated all fields
src/core/config/importExport.ts - Already has proper export logic
src/services/code-index/config-manager.ts - Correctly reads from nested structure
src/core/webview/webviewMessageHandler.ts - Correctly writes to nested structure
webview-ui/src/components/chat/CodeIndexPopover.tsx - Handles UI for settings
```

### Get in Touch

<!--
Please provide your Discord username for reviewers or maintainers to reach you if they have questions about your PR
-->

Discord: @contributor

daniel-lxs reviewed (approved):
LGTM
